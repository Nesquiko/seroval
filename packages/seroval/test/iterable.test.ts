import { describe, it, expect } from 'vitest';
import {
  compileJSON,
  deserialize,
  Feature,
  fromJSON,
  serialize,
  serializeAsync,
  toJSON,
  toJSONAsync,
} from '../src';

describe('Iterable', () => {
  describe('serialize', () => {
    it('supports Iterables', () => {
      const example = {
        title: 'Hello World',
        * [Symbol.iterator]() {
          yield 1;
          yield 2;
          yield 3;
        },
      };
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Iterable<number> & { title: string }>(result);
      expect(back.title).toBe(example.title);
      expect(Symbol.iterator in back).toBeTruthy();
      const iterator = back[Symbol.iterator]();
      expect(iterator.next().value).toBe(1);
      expect(iterator.next().value).toBe(2);
      expect(iterator.next().value).toBe(3);
    });
  });
  describe('serializeAsync', () => {
    it('supports Iterables', async () => {
      const example = Promise.resolve({
        title: 'Hello World',
        * [Symbol.iterator]() {
          yield 1;
          yield 2;
          yield 3;
        },
      });
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = await deserialize<typeof example>(result);
      expect(back.title).toBe((await example).title);
      expect(Symbol.iterator in back).toBeTruthy();
      const iterator = back[Symbol.iterator]();
      expect(iterator.next().value).toBe(1);
      expect(iterator.next().value).toBe(2);
      expect(iterator.next().value).toBe(3);
    });
  });
  describe('toJSON', () => {
    it('supports Iterables', () => {
      const example = {
        title: 'Hello World',
        * [Symbol.iterator]() {
          yield 1;
          yield 2;
          yield 3;
        },
      };
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Iterable<number> & { title: string }>(result);
      expect(back.title).toBe(example.title);
      expect(Symbol.iterator in back).toBeTruthy();
      const iterator = back[Symbol.iterator]();
      expect(iterator.next().value).toBe(1);
      expect(iterator.next().value).toBe(2);
      expect(iterator.next().value).toBe(3);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Iterables', async () => {
      const example = Promise.resolve({
        title: 'Hello World',
        * [Symbol.iterator]() {
          yield 1;
          yield 2;
          yield 3;
        },
      });
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<typeof example>(result);
      expect(back.title).toBe((await example).title);
      expect(Symbol.iterator in back).toBeTruthy();
      const iterator = back[Symbol.iterator]();
      expect(iterator.next().value).toBe(1);
      expect(iterator.next().value).toBe(2);
      expect(iterator.next().value).toBe(3);
    });
  });
  describe('compat', () => {
    it('should throw an error for unsupported target', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      expect(() => serialize(example, {
        disabledFeatures: Feature.Symbol,
      })).toThrowErrorMatchingSnapshot();
    });
    it('should use Symbol.iterator instead of Array.values.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      expect(serialize(example, {
        disabledFeatures: Feature.ArrayPrototypeValues,
      })).toMatchSnapshot();
    });
    it('should use method shorthand instead of arrow functions.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      expect(serialize(example, {
        disabledFeatures: Feature.ArrowFunction,
      })).toMatchSnapshot();
    });
    it('should use functions instead of method shorthand.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      expect(serialize(example, {
        disabledFeatures: Feature.MethodShorthand | Feature.ArrowFunction,
      })).toMatchSnapshot();
    });
  });
  describe('compat#toJSON', () => {
    it('should throw an error for unsupported target', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      expect(() => toJSON(example, {
        disabledFeatures: Feature.Symbol,
      })).toThrowErrorMatchingSnapshot();
    });
    it('should use Symbol.iterator instead of Array.values.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      const result = toJSON(example, {
        disabledFeatures: Feature.ArrayPrototypeValues,
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      expect(compileJSON(result)).toMatchSnapshot();
    });
    it('should use method shorthand instead of arrow functions.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      const result = toJSON(example, {
        disabledFeatures: Feature.ArrowFunction,
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      expect(compileJSON(result)).toMatchSnapshot();
    });
    it('should use functions instead of method shorthand.', () => {
      const example = {
        * [Symbol.iterator]() {
          yield example;
        },
      };
      const result = toJSON(example, {
        disabledFeatures: Feature.MethodShorthand | Feature.ArrowFunction,
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      expect(compileJSON(result)).toMatchSnapshot();
    });
  });
});
