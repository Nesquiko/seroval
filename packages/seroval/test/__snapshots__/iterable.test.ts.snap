// Vitest Snapshot v1

exports[`Iterable > compat > should throw an error for unsupported target 1`] = `"Unsupported type \\"Iterable\\""`;

exports[`Iterable > compat > should use Symbol.iterator instead of Array.values. 1`] = `"(h=>(h={[Symbol.iterator]:()=>[h][Symbol.iterator]()}))()"`;

exports[`Iterable > compat > should use functions instead of method shorthand. 1`] = `"(function(h){return h={[Symbol.iterator]:function(){return [h].values()}}})()"`;

exports[`Iterable > compat > should use method shorthand instead of arrow functions. 1`] = `"(function(h){return h={[Symbol.iterator](){return [h].values()}}})()"`;

exports[`Iterable > compat#toJSON > should throw an error for unsupported target 1`] = `"Unsupported type \\"Iterable\\""`;

exports[`Iterable > compat#toJSON > should use Symbol.iterator instead of Array.values. 1`] = `"{\\"t\\":{\\"t\\":21,\\"i\\":0,\\"l\\":1,\\"a\\":[{\\"t\\":10,\\"i\\":0}]},\\"r\\":0,\\"i\\":true,\\"f\\":8189,\\"m\\":[0]}"`;

exports[`Iterable > compat#toJSON > should use Symbol.iterator instead of Array.values. 2`] = `"(h=>(h={[Symbol.iterator]:()=>[h][Symbol.iterator]()}))()"`;

exports[`Iterable > compat#toJSON > should use functions instead of method shorthand. 1`] = `"{\\"t\\":{\\"t\\":21,\\"i\\":0,\\"l\\":1,\\"a\\":[{\\"t\\":10,\\"i\\":0}]},\\"r\\":0,\\"i\\":true,\\"f\\":8123,\\"m\\":[0]}"`;

exports[`Iterable > compat#toJSON > should use functions instead of method shorthand. 2`] = `"(function(h){return h={[Symbol.iterator]:function(){return [h].values()}}})()"`;

exports[`Iterable > compat#toJSON > should use method shorthand instead of arrow functions. 1`] = `"{\\"t\\":{\\"t\\":21,\\"i\\":0,\\"l\\":1,\\"a\\":[{\\"t\\":10,\\"i\\":0}]},\\"r\\":0,\\"i\\":true,\\"f\\":8187,\\"m\\":[0]}"`;

exports[`Iterable > compat#toJSON > should use method shorthand instead of arrow functions. 2`] = `"(function(h){return h={[Symbol.iterator](){return [h].values()}}})()"`;

exports[`Iterable > serialize > supports Iterables 1`] = `"(Object.assign({[Symbol.iterator]:()=>[1,2,3].values()},{title:\\"Hello World\\"}))"`;

exports[`Iterable > serializeAsync > supports Iterables 1`] = `"Promise.resolve(Object.assign({[Symbol.iterator]:()=>[1,2,3].values()},{title:\\"Hello World\\"}))"`;

exports[`Iterable > toJSON > supports Iterables 1`] = `"{\\"t\\":{\\"t\\":21,\\"i\\":0,\\"l\\":3,\\"d\\":{\\"k\\":[\\"title\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"Hello World\\"}],\\"s\\":1},\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]},\\"r\\":0,\\"i\\":true,\\"f\\":8191,\\"m\\":[]}"`;

exports[`Iterable > toJSONAsync > supports Iterables 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":21,\\"i\\":1,\\"l\\":3,\\"d\\":{\\"k\\":[\\"title\\"],\\"v\\":[{\\"t\\":1,\\"s\\":\\"Hello World\\"}],\\"s\\":1},\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;
